### PR TITLE
Make `ScalarWaveGr` accept `CurvedScalarWave` initial data directly

### PR DESCRIPTION
**Pull request by nikwit**
## Proposed changes

Currently the `ScalarWaveGr` class accepts only `ScalarWave` solutions and adjusts `Pi` to the background with spacetime quantities. There is no way to directly set `Pi` for the `CurvedScalarWave` initial data.

This PR adjusts `ScalarWaveGr` to recognize whether the wave solution//data comes from a `ScalarWave` solution where `Pi` needs to be adjusted with the background spacetime or directly from a `CurvedScalarWave` solution where `Pi` just needs to be forwarded.

- [x] depends on #3736

### CODE REVIEW

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp [hunk 1]
```diff
@@ -27,15 +28,21 @@ namespace AnalyticData {
...
+ * 2. solutions of the Einstein equations for the spacetime background.
+ *
+ * If the scalar field initial data returns `CurvedScalarWave` tags, \f$\Psi\f$,
+ * \f$\Pi\f$ and \f$\Phi_i\f$ will simply be forwarded from the initial data
+ * class. Alternatively, the scalar field initial data can be provided using any
+ * member class of `ScalarWave::Solutions` which return `ScalarWave` tags. In
+ * this case, \f$\Phi_i\f$ and \f$\Psi\f$ will also be forwarded but
+ * \f$\Pi\f$ will be adjusted to account for the curved background. Its
+ * definition comes from requiring it to be the future-directed time derivative
+ * of the scalar field in curved spacetime:
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:45]: Could you add a docstring to `CurvedScalarWave::Tags::Pi` with this definition in a separate commit, ideally pointing into a paper? It's missing for some reason. I think it's Eq. (14) in Holst2004wt (https://arxiv.org/abs/gr-qc/0407011).

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp [hunk 2]
```diff
@@ -88,6 +103,17 @@ class ScalarWaveGr : public MarkAsAnalyticData {
...
+  using InitialDataPsi =
+      tmpl::conditional_t<is_curved, CurvedScalarWave::Tags::Psi,
+                          ScalarWave::Tags::Psi>;
+  using InitialDataPi =
+      tmpl::conditional_t<is_curved, CurvedScalarWave::Tags::Pi,
+                          ScalarWave::Tags::Pi>;
+  using InitialDataPhi =
+      tmpl::conditional_t<is_curved, CurvedScalarWave::Tags::Phi<volume_dim>,
+                          ScalarWave::Tags::Phi<volume_dim>>;
+  using evolved_field_vars_tags =
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:115]: How about something like `scalar_wave_tags`? (squash directly if you change this name)

**Review by nilsvu — CHANGES_REQUESTED** [review]
LGTM. Note that we're moving to a single executable per dimension for each system, so once it's time to migrate the CurvedScalarWave executable this `ScalarWaveGr` class can factory-create the `ScalarField` and `Background` objects. At that point you'll probably have to rethink this implementation, since it's going to be a runtime choice. Still a few weeks off though.

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp [hunk 3]
```diff
@@ -53,13 +60,21 @@ class ScalarWaveGr : public MarkAsAnalyticData {
...
  public:
   static constexpr size_t volume_dim = ScalarFieldData::volume_dim;
-
+  static constexpr bool is_curved =
+      tmpl::list_contains_v<typename ScalarFieldData::tags,
+                            CurvedScalarWave::Tags::Psi>;
   struct ScalarField {
     using type = ScalarFieldData;
-    static constexpr Options::String help = {"Flat space scalar field."};
+    static constexpr Options::String help = {"Scalar field."};
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp]: [optional] Maybe expand to `The scalar field system data.` to make it clear it sets the subsystem?

File: src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/AnalyticData/CurvedWaveEquation/ScalarWaveGr.hpp:115]: I'm okay with leaving as-is or with renaming. It's unclear to me if there should be a preference here until we couple in the GH evolution.

File: tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp [hunk 4]
```diff
@@ -228,6 +243,36 @@ void test_construct_from_options() {
...
+            "  Radius: 10.\n"
+            "  Width: 1.\n"
+            "  Mode: [4, -3]");
+
+    // construct internals of data constructed from options
+    const gr::Solutions::KerrSchild bh_solution(0.7, {{0.1, -0.2, 0.3}},
+                                                {{0.7, 0.6, -2.}});
+    const CurvedScalarWave::AnalyticData::PureSphericalHarmonic wave_solution{
+        10., 1., {4, -3}};
+
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticData/CurvedWaveEquation/Test_ScalarWaveGr.cpp:267]: [optional] consider adding `const auto curved_scalar_wave_data = serialize_and_deserialize(..._from_opts);` and using that in the `CHECK` and the `test_kerr` call to verify serialization is working.

**Review by nilsdeppe — COMMENTED** [review]
overall LGTM, some suggestions you can squash in immediately if you like them

**Review by nilsvu — APPROVED** [review]
I think the `==` operator of `ScalarWaveGr` is wrong, please check in a follow-up PR.

**Comment by nikwit** [comment]
> I think the `==` operator of `ScalarWaveGr` is wrong, please check in a follow-up PR.

missing `scalar_wave_data_` comparison?

**Comment by nilsvu** [comment]
yep

**Review by nilsdeppe — APPROVED** [review]
LGTM! Let's defer the `operator==` since there might be issues with the ScalarWave solutions not having an `operator==` or something (I haven't checked). Some of those have `unique_ptr<BaseClass>` member variables

**Pull request merged**